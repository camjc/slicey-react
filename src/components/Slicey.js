/* @flow */
'use strict';

let React = require('react/addons');

require('styles/Slicey.scss');

type ArcDatum = {
  color: string;
  id: number;
  largeArcFlag: number;
  x1: number;
  x2: number;
  y1: number;
  y2: number;
};

type ArcData = Array<ArcDatum>

let SliceyArc = React.createClass({
  propTypes: {
    key: React.PropTypes.number,
    data: React.PropTypes.shape({
      color: React.PropTypes.string.isRequired,
      id: React.PropTypes.number.isRequired,
      largeArcFlag: React.PropTypes.number.isRequired,
      x1: React.PropTypes.number.isRequired,
      x2: React.PropTypes.number.isRequired,
      y1: React.PropTypes.number.isRequired,
      y2: React.PropTypes.number.isRequired,
    }),
  },
  render: function() {
    if (!this.props || !this.props.data) {return;}

    let data: ArcDatum = this.props.data;
    let d: string = `M0,0 L${data.x1},${data.y1} A0.5,0.5 0 ${data.largeArcFlag},1 ${data.x2},${data.y2} z`;
    return (
        <path className={data.color} d={d}></path>
      );
  },
});

const QUARTER: number = Math.PI / 2;
const HALF: number = Math.PI;
const ROUND: number = Math.PI * 2;
const RADIUS: number = 0.5;

type Dataset = [
  {
    status: string;
    value: number;
  }
]

let Slicey = React.createClass({
  propTypes: {
    dataset: React.PropTypes.arrayOf(
      React.PropTypes.shape({
        status: React.PropTypes.string.isRequired,
        value: React.PropTypes.number.isRequired,
      })),
    donut: React.PropTypes.bool.isRequired,
    diameter: React.PropTypes.number,
  },
  getPointX: function(angle: number): number {
    return RADIUS * Math.cos(angle);
  },

  getPointY: function(angle: number): number {
    return RADIUS * Math.sin(angle);
  },

  getArcs: function(dataset: Dataset): ArcData {

    let datum;
    let i;
    let len;
    let total = 0;
    for (i = 0, len = dataset.length; i < len; i++) {
      datum = dataset[i];
      total += datum.value;
    }

    let arcs = new Array(dataset.length);
    let endAngle = -QUARTER;
    let index;
    let j;
    let len1;
    let startAngle = 0;
    for (index = j = 0, len1 = dataset.length; j < len1; index = ++j) {
      datum = dataset[index];
      let angle = ROUND * datum.value / total;
      startAngle = endAngle;
      endAngle += angle;
      arcs[index] = {
        color: datum.status,
        id: index,
        largeArcFlag: angle > HALF ? 1 : 0,
        x1: this.getPointX(startAngle),
        x2: this.getPointX(endAngle),
        y1: this.getPointY(startAngle),
        y2: this.getPointY(endAngle),
      };
    }

    return arcs;
  },

  render: function() {
    if (!this.props || !this.props.dataset) {return;}

    let donut;
    if (this.props.donut) {
      donut = <circle className='donut' r='.25' cx='0' cy='0'></circle>;
    }

    let diameter: number = this.props.diameter || 100;
    let arcData: ArcData = this.getArcs(this.props.dataset);
    return (
        <svg className='Slicey' height={diameter} width={diameter} viewBox='-0.5 -0.5 1 1'>
          <circle className='background' r='.49' cx='0' cy='0'></circle>
          {arcData.map(function(arc: ArcDatum) {
            return <SliceyArc key={arc.id} data={arc}/>;
          })}
          {donut}
        </svg>
      );
  },
});

module.exports = Slicey;
