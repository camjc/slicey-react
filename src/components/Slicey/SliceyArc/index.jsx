/* eslint-disable import/no-extraneous-dependencies */
import PropTypes from 'prop-types';
import React from 'react';
import styles from '../styles.module.css';

const SliceyArc = ({
  color, x1, y1, largeArcFlag, x2, y2,
}) => (
  <path
    className={styles[color]}
    d={`M0,0 L${x1},${y1} A0.5,0.5 0 ${largeArcFlag},1 ${x2},${y2} z`}
  />
);

SliceyArc.propTypes = {
  color: PropTypes.string.isRequired,
  largeArcFlag: PropTypes.number.isRequired,
  x1: PropTypes.number.isRequired,
  x2: PropTypes.number.isRequired,
  y1: PropTypes.number.isRequired,
  y2: PropTypes.number.isRequired,
};

export default SliceyArc;
