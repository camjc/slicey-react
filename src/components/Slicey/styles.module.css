.slicey {
  vertical-align: top;
  margin: 30px;
}

.background {
  fill: #555;
}

.positive {
  fill: #95ac7c;
}

.negative {
  fill: #bc601d;
}

.info {
  fill: #bea552;
}

.warning,
.pending {
  fill: #efe6c0;
}

.donut {
  fill: #fff;
}
