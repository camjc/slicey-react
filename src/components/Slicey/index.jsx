/* eslint-disable import/no-extraneous-dependencies */
import PropTypes from 'prop-types';
import React from 'react';

import styles from './styles.module.css';
import SliceyArc from './SliceyArc';

const QUARTER = Math.PI / 2;
const HALF = Math.PI;
const ROUND = Math.PI * 2;
const RADIUS = 0.5;

const getPointX = angle => RADIUS * Math.cos(angle);

const getPointY = angle => RADIUS * Math.sin(angle);

const getArcs = (dataset) => {
  const total = dataset.reduce(
    (accumulator, { value }) => accumulator + value,
    0,
  );

  const arcs = new Array(dataset.length);
  let endAngle = -QUARTER;
  let startAngle = 0;

  dataset.forEach(({ value, status }, index) => {
    const angle = (ROUND * value) / total;
    startAngle = endAngle;
    endAngle += angle;
    arcs[index] = {
      color: status,
      key: index,
      largeArcFlag: angle > HALF ? 1 : 0,
      x1: getPointX(startAngle),
      x2: getPointX(endAngle),
      y1: getPointY(startAngle),
      y2: getPointY(endAngle),
    };
  });

  return arcs;
};

const Slicey = ({ dataset, hasDonut, diameter }) =>
  dataset && (
    <svg
      className={styles.slicey}
      height={diameter}
      width={diameter}
      viewBox="-0.5 -0.5 1 1"
    >
      <circle className={styles.background} r=".49" cx="0" cy="0" />
      {getArcs(dataset).map(arc => (
        <SliceyArc {...arc} />
      ))}
      {hasDonut && <circle className={styles.donut} r=".25" cx="0" cy="0" />}
    </svg>
  );

Slicey.propTypes = {
  dataset: PropTypes.arrayOf(PropTypes.shape({
    status: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
  })),
  hasDonut: PropTypes.bool,
  diameter: PropTypes.number,
};

Slicey.defaultProps = {
  dataset: null,
  diameter: 100,
  hasDonut: false,
};

export default Slicey;
