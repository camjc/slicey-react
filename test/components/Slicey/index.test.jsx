import React from 'react';
import { shallow, mount } from 'enzyme';

import Slicey from '../../../src/components/Slicey';

it('renders null without a dataset', () => {
  const wrapper = shallow(<Slicey />);
  expect(wrapper.equals(null)).toBe(true);
});

const dataset = [{ status: 'warning', value: 50 }, { status: 'error', value: 20 }];

it('renders svg with a dataset', () => {
  const wrapper = shallow(<Slicey dataset={dataset} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders svg with a dataset (mounted)', () => {
  const wrapper = mount(<Slicey dataset={dataset} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders a donut', () => {
  const wrapper = shallow(<Slicey dataset={dataset} hasDonut />);
  expect(wrapper.find('circle')).toHaveLength(2);
});
