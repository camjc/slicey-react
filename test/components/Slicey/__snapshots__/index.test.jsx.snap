// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders svg with a dataset (mounted) 1`] = `
<Slicey
  dataset={
    Array [
      Object {
        "status": "warning",
        "value": 50,
      },
      Object {
        "status": "error",
        "value": 20,
      },
    ]
  }
  diameter={100}
  hasDonut={false}
>
  <svg
    height={100}
    viewBox="-0.5 -0.5 1 1"
    width={100}
  >
    <circle
      cx="0"
      cy="0"
      r=".49"
    />
    <SliceyArc
      color="warning"
      key="0"
      largeArcFlag={1}
      x1={3.061616997868383e-17}
      x2={-0.4874639560909118}
      y1={-0.5}
      y2={0.11126046697815725}
    >
      <path
        d="M0,0 L3.061616997868383e-17,-0.5 A0.5,0.5 0 1,1 -0.4874639560909118,0.11126046697815725 z"
      />
    </SliceyArc>
    <SliceyArc
      color="error"
      key="1"
      largeArcFlag={0}
      x1={-0.4874639560909118}
      x2={-9.184850993605148e-17}
      y1={0.11126046697815725}
      y2={-0.5}
    >
      <path
        d="M0,0 L-0.4874639560909118,0.11126046697815725 A0.5,0.5 0 0,1 -9.184850993605148e-17,-0.5 z"
      />
    </SliceyArc>
  </svg>
</Slicey>
`;

exports[`renders svg with a dataset 1`] = `
<svg
  height={100}
  viewBox="-0.5 -0.5 1 1"
  width={100}
>
  <circle
    cx="0"
    cy="0"
    r=".49"
  />
  <SliceyArc
    color="warning"
    key="0"
    largeArcFlag={1}
    x1={3.061616997868383e-17}
    x2={-0.4874639560909118}
    y1={-0.5}
    y2={0.11126046697815725}
  />
  <SliceyArc
    color="error"
    key="1"
    largeArcFlag={0}
    x1={-0.4874639560909118}
    x2={-9.184850993605148e-17}
    y1={0.11126046697815725}
    y2={-0.5}
  />
</svg>
`;
